components/sso-hostobject/org.wso2.carbon.hostobjects.sso/src/main/java/org/wso2/carbon/hostobjects/sso/internal/util/Util.java

```diff
@@ -477,4 +480,34 @@ public static DocumentBuilderFactory getSecuredDocumentBuilder() {
 
         return dbf;
     }
+
+    /**
+     * Get the username from the SAML2 Assertion
+     *
+     * @param assertion SAML2 assertion
+     * @return username
+     */
+    public static String getUsernameFromAssertion(Assertion assertion, String usernameAttribute) {
+        if (usernameAttribute != null) {
+            // There can be multiple AttributeStatements in Assertion
+            List<AttributeStatement> attributeStatements = assertion.getAttributeStatements();
+            if (attributeStatements != null) {
+                for (AttributeStatement attributeStatement : attributeStatements) {
+                    // There can be multiple Attributes in an attributeStatement
+                    List<Attribute> attributes = attributeStatement.getAttributes();
+                    if (attributes != null) {
+                        for (Attribute attribute : attributes) {
+                            String attributeName = attribute.getDOM().getAttribute(SSOConstants.SAML_NAME_ATTRIBUTE);
+                            if (attributeName.equals(usernameAttribute)) {
+                                List<XMLObject> attributeValues = attribute.getAttributeValues();
+                                // There can be multiple attribute values in an attribute, but get the first one
+                                return attributeValues.get(0).getDOM().getTextContent();
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return assertion.getSubject().getNameID().getValue();
```

Comment:
we might need to check whether subject is null or not 

  Subject subject = assertions.get(0).getSubject();
 -                    if (subject != null) {
 -                        if (subject.getNameID() != null) {
 -                            username = subject.getNameID().getValue();
 -                            if (log.isDebugEnabled()) {
 -                                log.debug("Name of authenticated user from SAML response " + username);
 -                            }
 -                        }
 -                    }